pkg/reconciler/pipelinerun/cancel_test.go

```diff
@@ -104,46 +110,189 @@ func TestCancelPipelineRun(t *testing.T) {
 	}}
 	for _, tc := range testCases {
 		tc := tc
 		t.Run(tc.name, func(t *testing.T) {
+			var seedTaskRuns []*v1beta1.TaskRun
+			var seedRuns []*v1alpha1.Run
+
+			for _, tr := range tc.taskRuns {
+				seedTaskRuns = append(seedTaskRuns, &v1beta1.TaskRun{
+					ObjectMeta: metav1.ObjectMeta{Name: tr.Name,
+						Labels: map[string]string{"shouldCancel": "true"}},
+				})
+			}
+			for _, tr := range tc.ignoredTaskRuns {
+				seedTaskRuns = append(seedTaskRuns, &v1beta1.TaskRun{
+					ObjectMeta: metav1.ObjectMeta{Name: tr.Name,
+						Labels: map[string]string{"shouldCancel": "false"}},
+				})
+			}
+
+			for _, r := range tc.runs {
+				seedRuns = append(seedRuns, &v1alpha1.Run{
+					ObjectMeta: metav1.ObjectMeta{Name: r.Name,
+						Labels: map[string]string{"shouldCancel": "true"}},
+				})
+			}
+			for _, r := range tc.ignoredRuns {
+				seedRuns = append(seedRuns, &v1alpha1.Run{
+					ObjectMeta: metav1.ObjectMeta{Name: r.Name,
+						Labels: map[string]string{"shouldCancel": "false"}},
+				})
+			}
+
 			d := test.Data{
 				PipelineRuns: []*v1beta1.PipelineRun{tc.pipelineRun},
-				TaskRuns:     tc.taskRuns,
-				Runs:         tc.runs,
+				TaskRuns:     seedTaskRuns,
+				Runs:         seedRuns,
 			}
 			ctx, _ := ttesting.SetupFakeContext(t)
 			ctx, cancel := context.WithCancel(ctx)
 			defer cancel()
 			c, _ := test.SeedTestData(t, ctx, d)
-			if err := cancelPipelineRun(ctx, logtesting.TestLogger(t), tc.pipelineRun, c.Pipeline); err != nil {
-				t.Fatal(err)
-			}
-			// This PipelineRun should still be complete and false, and the status should reflect that
-			cond := tc.pipelineRun.Status.GetCondition(apis.ConditionSucceeded)
-			if cond.IsTrue() {
-				t.Errorf("Expected PipelineRun status to be complete and false, but was %v", cond)
-			}
-			if tc.taskRuns != nil {
-				l, err := c.Pipeline.TektonV1beta1().TaskRuns("").List(ctx, metav1.ListOptions{})
+
+			err := cancelPipelineRun(ctx, logtesting.TestLogger(t), tc.pipelineRun, c.Pipeline)
+			if tc.expectedErr != nil {
+				if err == nil {
+					t.Fatalf("expected error '%s', but did not get an error", tc.expectedErr)
+				}
+				if tc.expectedErr.Error() != err.Error() {
+					t.Fatalf("expected error '%s', but got '%s'", tc.expectedErr, err)
+				}
+			} else {
 				if err != nil {
 					t.Fatal(err)
 				}
-				for _, tr := range l.Items {
-					if tr.Spec.Status != v1beta1.TaskRunSpecStatusCancelled {
-						t.Errorf("expected task %q to be marked as cancelled, was %q", tr.Name, tr.Spec.Status)
+				// This PipelineRun should still be complete and false, and the status should reflect that
+				cond := tc.pipelineRun.Status.GetCondition(apis.ConditionSucceeded)
+				if cond.IsTrue() {
+					t.Errorf("Expected PipelineRun status to be complete and false, but was %v", cond)
+				}
+				if tc.taskRuns != nil {
+					l, err := c.Pipeline.TektonV1beta1().TaskRuns("").List(ctx, metav1.ListOptions{
```

Comment:
listing by label selector is a bit roundabout, could you just query by the names of the ignored/not ignored runs?

Reply:
I can't remember my reasoning for this approach off the top of my head, but I do remember that this felt like the clearest way to get what we needed.

Reply:
Is this approach used elsewhere in the repo? Do you think querying by name could work?